### PR TITLE
feat: LT-1400 - Exporter Roles and Permissions

### PR DESCRIPTION
**Pull request by cstobart**

### CODE REVIEW

File: core/views.py [hunk 1]
```diff
@@ -14,6 +15,19 @@
...
+            manage_organisation_section_link = reverse_lazy("users:users")
+            title = "Manage my users"
+        elif Permissions.ADMINISTER_SITES in user_permissions:
+            manage_organisation_section_link = reverse_lazy("sites:sites")
+            title = "Manage my sites"
+        elif Permissions.EXPORTER_ADMINISTER_ROLES in user_permissions:
+            manage_organisation_section_link = reverse_lazy("roles:roles")
+            title = "Manage my roles"
+        else:
+            manage_organisation_section_link = None
```
Comment (Mark-Spiers-kainos) [inline core/views.py]: Would there be a situation where a user might have multiple permissions listed above?
If so how is this handled?

File: roles/admin.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from django.contrib import admin
+
+# Register your models here.
```
Comment (Mark-Spiers-kainos) [inline roles/admin.py]: not sure if this file is needed, or roles/models

File: core/views.py (hunk 1 above)
Comment (cstobart) [reply core/views.py]: line 28

File: roles/admin.py (hunk 2 above)
Comment (andytaylorkainos) [inline roles/admin.py]: do we need this file?

File: roles/apps.py [hunk 3]
```diff
@@ -0,0 +1,5 @@
+from django.apps import AppConfig
```
Comment (andytaylorkainos) [inline roles/apps.py]: do we actually need these files?

File: roles/forms.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+from django.http import HttpRequest
```
Comment (andytaylorkainos) [inline roles/forms.py]: content for file pls :)

File: roles/models.py [hunk 5]
```diff
@@ -0,0 +1,3 @@
+from django.db import models
```
Comment (andytaylorkainos) [inline roles/models.py]: unused file

File: roles/services.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+
+
+def post_sites(request, organisation_id, json):
+    data = post(request, ORGANISATIONS_URL + organisation_id + SITES_URL, json)
+    return data.json(), data.status_code
+
+
+def get_roles(request, organisation_id, convert_to_options=False):
+    data = get(request, ORGANISATIONS_URL + str(organisation_id) + ROLES_URL)
+    print(data)
```
Comment (andytaylorkainos) [inline roles/services.py]: print statement

File: roles/services.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+            converted.append(Option(key=item["id"], value=item["name"]))
+
+        return converted
+
+    return data.json()["permissions"]
+
+
+def get_user_permissions(request):
+    user, _ = get_user(request)
+    print(user)
```
Comment (andytaylorkainos) [inline roles/services.py]: print statement

File: core/helpers.py [hunk 8]
```diff
@@ -111,3 +111,10 @@ def remove_prefix(json, prefix):
             field = k[len(prefix):]
             post_data[field] = json[k]
     return post_data
+
+
+def println(content=None, no=1):
```
[the remark this answers is not in the document]
Comment (cstobart) [reply core/helpers.py]: its a dev tool, really very helpful

File: applications/libraries/task_lists.py [hunk 9]
```diff
@@ -28,6 +29,9 @@ def _get_standard_application_task_list(request, application, errors=None):
     is_editing = False
     edit_type = None
 
+    user_permissions = get_user_permissions(request)
+    submit = True if Permissions.SUBMIT_LICENCE_APPLICATION in user_permissions else False
```
[the remark this answers is not in the document]
Comment (cstobart) [reply applications/libraries/task_lists.py]: 👍

File: roles/views.py [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        self.form = add_role(request)
+        self.action = post_role
+        self.success_url = reverse_lazy("roles:roles")
+
+
+class EditRole(SingleFormView):
+    def init(self, request, **kwargs):
+        self.object_pk = kwargs["pk"]
+        self.data = get_role(request, self.object_pk)
+        println(self.data)
```
Comment (jfaracik) [inline roles/views.py]: Println here!

File: roles/services.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def put_role(request, pk, json):
+    organisation_id = request.user.organisation
+    data = put(request, ORGANISATIONS_URL + str(organisation_id) + ROLES_URL + str(pk) + "/", json)
+    return data.json(), data.status_code
+
+
+def get_permissions(request, convert_to_options=False):
+    data = get(request, EXPORTER_USERS_PERMISSIONS_URL)
```
Comment (jfaracik) [inline roles/services.py]: .json()['permissions'] at the end of this line, then you won't need to duplicate ['permissions'] elsewhere in the function

File: roles/urls.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+from django.urls import path
+
+from . import views
```
Comment (jfaracik) [inline roles/urls.py]: `from roles import views` not `.`

File: roles/urls.py [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+
+app_name = "roles"
+urlpatterns = [
+    # ex: /sites/
```
Comment (jfaracik) [inline roles/urls.py]: Delete comments

File: roles/services.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+
+def post_sites(request, organisation_id, json):
+    data = post(request, ORGANISATIONS_URL + organisation_id + SITES_URL, json)
+    return data.json(), data.status_code
+
+
+def get_roles(request, organisation_id, convert_to_options=False):
+    data = get(request, ORGANISATIONS_URL + str(organisation_id) + ROLES_URL)
+    data = data.json()["results"]
```
Comment (jfaracik) [inline roles/services.py]: Make this one line.

File: roles/services.py [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+    if convert_to_options:
+        converted = []
+
+        for item in data.json().get("roles"):
```
Comment (jfaracik) [inline roles/services.py]: Then change this accordingly

File: templates/roles/index.html [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+{% block back_link %}
+	{% include 'includes/breadcrumbs.html' %}
+{% endblock %}
+
+{% block body %}
+
+	<h2 class="govuk-heading-l">{{ organisation.name }}</h2>
+
+	<div style="display: grid; grid-template-columns: 1fr auto; grid-gap: 30px;">
+		<div class="lite-tabs--fade">
```
Comment (jfaracik) [inline templates/roles/index.html]: Can you change `--fade` to `__container`

https://lite-patterns.london.cloudapps.digital/components/tabs

File: ui_automation_tests/pages/shared.py [hunk 17]
```diff
@@ -6,16 +6,18 @@ class Shared(BasePage):
...
     GOV_TABLE = ".govuk-table"
     GOV_TABLE_ROW = ".govuk-table__row"
     GOV_TABLE_CELL_LINKS = ".govuk-table__cell a"
+    GOV_TABLE_CELL = ".govuk-table__cell"
     APPLICATION_NAME = "a[href*='overview']"
     H2 = "h2"
     H1 = "h1"
     LITE_SECTION = ".lite-section"
     NOTIFICATION = ".app-icon-label__notification"  # CSS
+    SUBMIT_BUTTON = '.govuk-button[type*="submit"]'
```
Comment (jfaracik) [inline ui_automation_tests/pages/shared.py]: Remove this, use the function functions.click_submit()

File: ui_automation_tests/pages/shared.py [hunk 18]
```diff
@@ -70,3 +72,9 @@ def get_size_of_table_rows(self):
 
     def click_on_application_name(self):
         return self.driver.find_element_by_css_selector(self.APPLICATION_NAME).click()
+
+    def get_cells_in_gov_table(self):
+        return self.driver.find_elements_by_css_selector(self.GOV_TABLE_CELL)
+
+    def click_submit(self):
```
Comment (jfaracik) [inline ui_automation_tests/pages/shared.py]: Remove this, use the function functions.click_submit()

File: ui_automation_tests/step_defs/test_roles.py [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+    if role_name == " ":
+        context.role_name = role_name
+    else:
+        extra_string = str(utils.get_unformatted_date_time())
+        extra_string = extra_string[(len(extra_string)) - 14 :]
+        context.role_name = role_name + extra_string
+
+    roles_page.enter_role_name(context.role_name)
+    roles_page.select_permissions(permissions)
+    Shared(driver).click_submit()
```
Comment (jfaracik) [inline ui_automation_tests/step_defs/test_roles.py]: Remove this, use the function functions.click_submit() - You may need to pull the latest ui_automation_test/shared

File: ui_automation_tests/pages/shared.py [hunk 20]
```diff
@@ -6,11 +6,12 @@ class Shared(BasePage):
     RADIO_BUTTONS = ".govuk-radios__label"  # CSS
     ERROR_MESSAGES = ".govuk-error-summary__body"
     ERROR_MESSAGE = ".govuk-error-message"
-    GOV_BODY = ".govuk-body"
+    GOV_BODY = ".govuk-table__body"
```
Comment (jfaracik) [inline ui_automation_tests/pages/shared.py]: This should be GOV_TABLE_BODY

File: ui_automation_tests/pages/roles_pages.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
+from shared.BasePage import BasePage
+
+
+class RolesPages(BasePage):
+
+    ADD_ROLE_TEXT_FIELD = "name"  # ID
+    ADD_ROLE = "button_add_role"  # ID
+    MANAGE_ROLES_BUTTON = 'a[href="/roles/"]'
```
Comment (jfaracik) [inline ui_automation_tests/pages/roles_pages.py]: Make this an id, eg `button-roles'

File: templates/users/profile.html [hunk 22]
```diff
@@ -1,5 +1,9 @@
 {% extends 'layouts/two_pane.html' %}
 
+{% block back_link %}
+    <a href="{% url 'users:users' %}" id="back-link" class="govuk-back-link">Back to Users</a>
```
Comment (andytaylorkainos) [inline templates/users/profile.html]: String here needs to go in content

File: templates/applications/submit.html [hunk 23]
```diff
@@ -26,7 +26,8 @@ <h1 class="govuk-heading-xl">
 			<label class="govuk-label govuk-checkboxes__label" for="declaration">The information I have given on this application is correct</label>
 		</div>
 	</div>
-
-	<a href="confirmation.html" class="govuk-button">Submit application</a>
+    {% if can_submit %}
+	    <a href="confirmation.html" class="govuk-button">Submit application</a>
```
Comment (andytaylorkainos) [inline templates/applications/submit.html]: String here needs to go in content

File: templates/applications/standard-application-edit.html [hunk 24]
```diff
@@ -517,26 +517,29 @@ <h2 class="lite-task-list__section">
...
-							{% else %}
-								<a href="{% url 'applications:application' application.id %}" class="govuk-button" draggable="false" role="button">Done</a>
+				{% if can_submit %}
+					<li>
+						<div class="app-submit-bar">
+							<form method="post">
+								{% csrf_token %}
+								{% if not application.status.value or application.status.key == 'applicant_editing' %}
+									<button type="submit" value="submit" class="govuk-button" data-prevent-double-click="true" style="margin-bottom: 0;">Submit application</button>
+								{% else %}
```
Comment (andytaylorkainos) [inline templates/applications/standard-application-edit.html]: String here needs to go in content

File: roles/forms.py (hunk 4 above)
Comment (andytaylorkainos) [reply roles/forms.py]: I think all these titles/questions need to go into content :(

**Pull request merged**